### PR TITLE
perf(ci, emulator): community emulator action / AVD caching / gradle caching

### LINKED ISSUE #10601: [Bug] Emulator tests failing: TimeoutCancellationException
`kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 120000 ms`

```
com.google.testing.platform.api.plugin.PluginException: Exception thrown during onBeforeAll invocation of plugin com.google.testing.platform.plugin.android.AndroidDevicePlugin.
Test results saved as file:/Users/runner/work/Anki-Android/Anki-Android/AnkiDroid/build/outputs/androidTest-results/connected/flavors/play/TestingAVD(AVD)%20-%2011/test-result.pb. Inspect these results in Android Studio by selecting Run > Import Tests From File from the menu bar and importing test-result.pb.
	at com.google.testing.platform.plugin.PluginLifecycleKt.invokeOrThrow(PluginLifecycle.kt:216)
	at com.google.testing.platform.plugin.PluginLifecycleKt.invokeOrThrow$default(PluginLifecycle.kt:205)
	at com.google.testing.platform.plugin.PluginLifecycle$onBeforeAll$1.invoke(PluginLifecycle.kt:92)
	at com.google.testing.platform.plugin.PluginLifecycle$onBeforeAll$1.invoke(PluginLifecycle.kt:88)
	at com.google.testing.platform.core.telemetry.common.noop.NoopDiagnosticsScope.recordEvent(NoopDiagnosticsScope.kt:35)
	at com.google.testing.platform.core.telemetry.TelemetryKt.recordEvent(Telemetry.kt:105)
	at com.google.testing.platform.core.telemetry.TelemetryKt.recordEvent$default(Telemetry.kt:98)
	at com.google.testing.platform.plugin.PluginLifecycle.onBeforeAll(PluginLifecycle.kt:88)
	at com.google.testing.platform.executor.SingleDeviceExecutor$execute$4.invoke(SingleDeviceExecutor.kt:86)
	at com.google.testing.platform.executor.SingleDeviceExecutor$execute$4.invoke(SingleDeviceExecutor.kt:86)
	at com.google.testing.platform.executor.SingleDeviceExecutor.runUnlessCancelled(SingleDeviceExecutor.kt:105)
	at com.google.testing.platform.executor.SingleDeviceExecutor.execute(SingleDeviceExecutor.kt:86)
	at com.google.testing.platform.RunnerImpl.run(RunnerImpl.kt:108)
	at com.google.testing.platform.server.strategy.NonInteractiveServerStrategy$run$4.invoke(NonInteractiveServerStrategy.kt:80)
	at com.google.testing.platform.server.strategy.NonInteractiveServerStrategy$run$4.invoke(NonInteractiveServerStrategy.kt:79)
	at com.google.testing.platform.core.telemetry.common.noop.NoopDiagnosticsScope.recordEvent(NoopDiagnosticsScope.kt:35)
	at com.google.testing.platform.core.telemetry.TelemetryKt.recordEvent(Telemetry.kt:66)
	at com.google.testing.platform.server.strategy.NonInteractiveServerStrategy.run(NonInteractiveServerStrategy.kt:79)
	at com.google.testing.platform.main.MainKt$main$4.invokeSuspend(Main.kt:67)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at com.google.testing.platform.main.MainKt.main(Main.kt:66)
	at com.google.testing.platform.main.MainKt.main$default(Main.kt:34)
	at com.google.testing.platform.main.MainKt.main(Main.kt)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.google.testing.platform.launcher.Launcher.main(Launcher.java:149)
Caused by: kotlinx.coroutines.TimeoutCancellationException: Timed out waiting for 120000 ms
	at kotlinx.coroutines.TimeoutKt.TimeoutCancellationException(Timeout.kt:158)
	at kotlinx.coroutines.TimeoutCoroutine.run(Timeout.kt:128)
	at kotlinx.coroutines.EventLoopImplBase$DelayedRunnableTask.run(EventLoop.common.kt:497)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.DefaultExecutor.run(DefaultExecutor.kt:68)
	at java.base/java.lang.Thread.run(Thread.java:829)

Mar 21, 2022 11:24:49 PM com.android.tools.utp.plugins.host.additionaltestoutput.AndroidAdditionalTestOutputPlugin deviceShellAndCheckSuccess
WARNING: Shell command failed (1): ls "/sdcard/Android/media/com.ichi2.anki/additional_test_output"
ls: /sdcard/Android/media/com.ichi2.anki/additional_test_output: No such file or directory

Mar 21, 2022 11:24:49 PM com.android.tools.utp.plugins.host.additionaltestoutput.AndroidAdditionalTestOutputPlugin afterAll
WARNING: Failed to retrieve additional test outputs from device.
com.android.ddmlib.SyncException: Remote object doesn't exist!
	at com.android.ddmlib.SyncService.pullFile(SyncService.java:341)
	at com.android.ddmlib.internal.DeviceImpl.pullFile(DeviceImpl.java:1083)
	at com.android.tools.utp.plugins.deviceprovider.ddmlib.DdmlibAndroidDevice.pullFile(DdmlibAndroidDevice.kt)
	at com.android.tools.utp.plugins.deviceprovider.ddmlib.DdmlibAndroidDeviceController.pull(DdmlibAndroidDeviceController.kt:254)
	at com.google.testing.platform.executor.DeviceControllerProxy$pull$2.invoke(DeviceControllerProxy.kt:82)
	at com.google.testing.platform.executor.DeviceControllerProxy$pull$2.invoke(DeviceControllerProxy.kt:82)
	at com.google.testing.platform.core.telemetry.common.noop.NoopDiagnosticsScope.recordEvent(NoopDiagnosticsScope.kt:35)
	at com.google.testing.platform.core.telemetry.TelemetryKt.recordEvent(Telemetry.kt:66)
	at com.google.testing.platform.executor.DeviceControllerProxy.pull(DeviceControllerProxy.kt:79)
	at com.android.tools.utp.plugins.host.additionaltestoutput.AndroidAdditionalTestOutputPlugin.copyAdditionalTestOutputsFromDeviceToHost(AndroidAdditionalTestOutputPlugin.kt:222)
	at com.android.tools.utp.plugins.host.additionaltestoutput.AndroidAdditionalTestOutputPlugin.afterAll(AndroidAdditionalTestOutputPlugin.kt:206)
	at com.google.testing.platform.plugin.PluginLifecycle$onAfterAll$1$2$1.invoke(PluginLifecycle.kt:180)
	at com.google.testing.platform.plugin.PluginLifecycle$onAfterAll$1$2$1.invoke(PluginLifecycle.kt:179)
	at com.google.testing.platform.plugin.PluginLifecycleKt$invokeOrThrow$1.invoke(PluginLifecycle.kt:213)
	at com.google.testing.platform.plugin.PluginLifecycleKt$invokeOrThrow$1.invoke(PluginLifecycle.kt:212)
	at com.google.testing.platform.core.telemetry.common.noop.NoopDiagnosticsScope.recordEvent(NoopDiagnosticsScope.kt:35)
	at com.google.testing.platform.core.telemetry.TelemetryKt.recordEvent(Telemetry.kt:105)
	at com.google.testing.platform.core.telemetry.TelemetryKt.recordEvent$default(Telemetry.kt:98)
	at com.google.testing.platform.plugin.PluginLifecycleKt.invokeOrThrow(PluginLifecycle.kt:212)
	at com.google.testing.platform.plugin.PluginLifecycleKt.invokeOrThrow$default(PluginLifecycle.kt:205)
	at com.google.testing.platform.plugin.PluginLifecycle$onAfterAll$1.invoke(PluginLifecycle.kt:179)
	at com.google.testing.platform.plugin.PluginLifecycle$onAfterAll$1.invoke(PluginLifecycle.kt:173)
	at com.google.testing.platform.core.telemetry.common.noop.NoopDiagnosticsScope.recordEvent(NoopDiagnosticsScope.kt:35)
	at com.google.testing.platform.core.telemetry.TelemetryKt.recordEvent(Telemetry.kt:105)
	at com.google.testing.platform.core.telemetry.TelemetryKt.recordEvent$default(Telemetry.kt:98)
	at com.google.testing.platform.plugin.PluginLifecycle.onAfterAll(PluginLifecycle.kt:173)
	at com.google.testing.platform.executor.SingleDeviceExecutor$execute$3.invoke(SingleDeviceExecutor.kt:83)
	at com.google.testing.platform.executor.SingleDeviceExecutor$execute$3.invoke(SingleDeviceExecutor.kt:82)
	at com.google.testing.platform.result.TestResultListenerManager.afterTestSuite(TestResultListenerManager.kt:[140](https://github.com/ankidroid/Anki-Android/runs/5636084762?check_suite_focus=true#step:14:140))
	at com.google.testing.platform.RunnerImpl.run(RunnerImpl.kt:118)
	at com.google.testing.platform.server.strategy.NonInteractiveServerStrategy$run$4.invoke(NonInteractiveServerStrategy.kt:80)
	at com.google.testing.platform.server.strategy.NonInteractiveServerStrategy$run$4.invoke(NonInteractiveServerStrategy.kt:79)
	at com.google.testing.platform.core.telemetry.common.noop.NoopDiagnosticsScope.recordEvent(NoopDiagnosticsScope.kt:35)
	at com.google.testing.platform.core.telemetry.TelemetryKt.recordEvent(Telemetry.kt:66)
	at com.google.testing.platform.server.strategy.NonInteractiveServerStrategy.run(NonInteractiveServerStrategy.kt:79)
	at com.google.testing.platform.main.MainKt$main$4.invokeSuspend(Main.kt:67)
	at kotlin.coroutines.jvm.internal.BaseContinuationImpl.resumeWith(ContinuationImpl.kt:33)
	at kotlinx.coroutines.DispatchedTask.run(DispatchedTask.kt:106)
	at kotlinx.coroutines.EventLoopImplBase.processNextEvent(EventLoop.common.kt:274)
	at kotlinx.coroutines.BlockingCoroutine.joinBlocking(Builders.kt:85)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking(Builders.kt:59)
	at kotlinx.coroutines.BuildersKt.runBlocking(Unknown Source)
	at kotlinx.coroutines.BuildersKt__BuildersKt.runBlocking$default(Builders.kt:38)
	at kotlinx.coroutines.BuildersKt.runBlocking$default(Unknown Source)
	at com.google.testing.platform.main.MainKt.main(Main.kt:66)
	at com.google.testing.platform.main.MainKt.main$default(Main.kt:34)
	at com.google.testing.platform.main.MainKt.main(Main.kt)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at com.google.testing.platform.launcher.Launcher.main(Launcher.java:[149](https://github.com/ankidroid/Anki-Android/runs/5636084762?check_suite_focus=true#step:14:149))


Platform error occurred when running the UTP test suite
```

StackOverflow implies that the device hasn't booted. It might be a worthwhile experiment (in a branch) to sleep for 5 minutes after `sys.boot_completed` to confirm if this diagnosis is correct.

https://stackoverflow.com/questions/71529420/android-instrumentation-test-fails-to-install-apk

### LINKED ISSUE #10672: Warn that emulator tests are flaky
Some new contributors see emulator tests failing and don't get why. We have no automated way to let them know this test is flaky.

So I'd suggest that if the emulator tests fail, a message should appear stating something such as:
> Emulator tests are currently flaky. If you did not do change emulator tests, you can probably ignore the failure; a maintainer will restart the test or let you know if you actually need to make some change to your PR.

Ideally, it would be nice to have a simple way to activate/deactivate it, a parameter somewhere. This way, if one day we get to ensure the tests are not flaky (or at least, far less flaky), we can deactivate it. And reactivate the day they are become flaky again.

An extra feature would be not to have this message if the submitter is in the group of maintainer, as we can safely assume we all know the state of emulator test.

I suppose you can do this either with a bot, or with an extra step of test automation. But honestly, I don't know github enough to know how to do it

### PR DESCRIPTION
**Pull request by mikehardy**
## Purpose / Description

- we have lots of flakes right now #10672 / #10601 
- we no longer have a multi-API test matrix for emulators, we used to have it on travis
- CI can be a bit slow

This PR updates our CI emulator infrastructure to use a community emulator runner that everyone else uses, to do caching a lot more strictly so that branches can't evict our main cache as easily, to cache emulator snapshots and gradle a bit better for speed, and to run on our lowest and highest supported APIs

## Fixes
Fixes #10672 (by making it not necessary to warn, it shouldn't flake now)
Fixes #10601 by making it not flake

## Approach
- switch from my custom mostly anti-flake until recently emu startup stuff to react-circus's community-choice android runner
- carefully enable caching for the AVD on main branch
- switch to community consensus gradle runner to get better caching there, but only on main branch
- choose more precise gradle targets for the gradle warmup pre-emu
- run the whole thing in a matrix

## How Has This Been Tested?

So many iterations....all on my fork. Here's the one where I was happy with hit

https://github.com/mikehardy/Anki-Android/actions/runs/2209842877

This is an uncached version of it / cold emu boot / no cache and it's got no consistent performance regression vs our current emulator run

When I run it with caches on the branch so you get the warm behavior, it shaves an inconsistent amount off CI but maybe 5 minutes off 20-25?

## Learning (optional, can help others)

I spent a fair amount of time thinking about cache eviction for performance:
- github actions now allows 10GB of space
- caches are not shared among any branches except that branches may have read access to main branch cache
- caches are LRU evicted

So the way we do branches and builds and stuff I think our cache performance can't be good though - cardinal sin! - I did not investigate hit ratios. I just know that branches must be evicting off main cache entries because gradle is ~700MB of cache, and now with 2 emulator images of 1GB a piece that would mean just 12 branch builds between a main build would do it before caching emulators, and now only 4 would do it.

Now main's cache will be mostly up to date for builds, especially with emu startup and branches will benefit.

I predict much shorter single run times on main and a large improvement in average time for all runs.

## Checklist
_Please, go through these checks before submitting the PR._

- [x] You have not changed whitespace unnecessarily (it makes diffs hard to read)
- [x] You have a descriptive commit message with a short title (first line, max 50 chars).
- [x] Your code follows the style of the project (e.g. never omit braces in `if` statements) 
- [x] You have commented your code, particularly in hard-to-understand areas
- [x] You have performed a self-review of your own code
- [ ] UI changes: include screenshots of all affected screens (in particular showing any new or changed strings)
- [ ] UI Changes: You have tested your change using the [Google Accessibility Scanner](https://play.google.com/store/apps/details?id=com.google.android.apps.accessibility.auditor)

### CODE REVIEW

**Comment by mikehardy** [comment]
it's a need review for interested people --> pending merge as default stance if no comments in 24ish hours
I'll pick up the pieces obviously if I merge and anything goes wrong, the CI is my baby :)

**Comment by david-allison** [comment]
Just to note: the linked 'happy with it' failed with a timeout: https://github.com/mikehardy/Anki-Android/runs/6134561484?check_suite_focus=true

File: .github/workflows/tests_emulator.yml [hunk 1]
```diff
@@ -37,108 +37,81 @@ jobs:
     name: Android Emulator Test
     runs-on: macos-latest
     timeout-minutes: 60
-    env:
-      EMULATOR_COMMAND: "-avd TestingAVD -noaudio -gpu swiftshader_indirect -camera-back none -no-snapshot -no-window -no-boot-anim -nojni -memory 2048 -timezone 'Europe/London' -cores 2"
-      EMULATOR_EXECUTABLE: qemu-system-x86_64-headless
+    strategy:
+      fail-fast: false
```
Comment (david-allison) [inline .github/workflows/tests_emulator.yml:41]: What impact does this have on coverage?

**Review by david-allison — COMMENTED** [review]
LGTM, thanks so much! One question

**Comment by david-allison** [comment]
> Emulator Tests / Android Emulator Test (21) (pull_request) Successful in 20m
> Emulator Tests / Android Emulator Test (32) (pull_request) Successful in 36m

Seems like a performance regression

**Comment by mikehardy** [comment]
Aggregating response:

- the linked happy with it was going to succeed and I increased the timeout in a last push here because 31 and 32 are slower
- the performance regression is because we're using a different emu that is slower but also represents our top API supported, I feel like that's important, it's comparable in the cold case on same API (30) and faster when warm
- What impact does matrix strategy fail-fast false have on coverage? well in general with multiple APIs covered we should get more compat coverage though robolectric handles that, specifically if one fails but another doesn't, we'll get cover shown for a full emulator (of the non-failed API) but not the other. Should be miniscule

Thoughts pro:

- I think the warm case is going to be a little faster than our current, and most everyone will be "warm" on this after first `main` run
- I think the cold case is going to be just a touch slower with this because I put a 3-minute sleep in on the cold case to make sure the snapshot was truly warm before taking it, but no one should hit the cold case after first snapshot save
- I think the warm case would be a *lot* faster if we were using the fastest AVDs but I really want a top/bottom API level bracket and that argues for 32 which is a bit slow

Bottom line: cold case a touch slower, warm case a bit faster, more confidence via API support bracket and no flakes :crossed_fingers:

**Comment by mikehardy** [comment]
https://github.com/ankidroid/Anki-Android/runs/6134931509?check_suite_focus=true

16 minutes (8 gradle build + 8 emu install + boot) should go away from the 32mins of API32, replaced by ~1min of cache unpack+utilize, leaving it with 32-15 or around 17mins execution

API32 executes tests in ~8mins
API21 executes tests in ~0mins because :astonished:  it is not actually discovering + running tests

https://github.com/mikehardy/Anki-Android/runs/6134561425?check_suite_focus=true#step:8:406

Okay, I'm still comfortable with all of the PR work here *except* for androidTest discovery on this API21 emu. Going to go waiting for feedback until I get an emu that actually runs tests, either different flavor here or higher API. :weary:

**Comment by david-allison** [comment]
Understood. Good to go on this

> the performance regression is because we're using a different emu that is slower but also represents our top API supported

Is it feasible to use the 'faster' emulator for the lower API - this means we're not holding as many macOS runners.

After this PR: we need 3 macOS runners per job (Unit test, and 2x Android Test). GitHub allocate us 5 concurrent macOS runners.

Holding up these resources /could/ cause a stall in the PR pipeline (we've gone from 5 concurrent PRs to 1.6...) and speeding up one runner could avoid this issue.

**Comment by mikehardy** [comment]
may have no choice but to use the faster one for the lower API if I can't get test discovery going on the really old APIs, I'll see what I can do.
Definitely sensitive to pipeline stuffage
I'm going to play with the caching on the unit test ones as well as I think builds will speed up in general with better caching now that I understand how all the caching we do is likely LRU-evicting main cache, which is then making branch cache cold as well

**Comment by mikehardy** [comment]
Okay I'll update it to use a "reasonable" (not too slow... no flakes) bracket of APIs based on the results from my massive run from the full emulator performance test suite I just built

https://gist.github.com/mikehardy/f076172f3b28826898c55dc03fe202de

image

**Comment by david-allison** [comment]
That's mindblowingly awesome!!!!

**Comment by mikehardy** [comment]
Unfortunately I'm not really satisfied with what I think as "reasonable" for frequent execution from a performance perspective.

Here's what I'm going to do: keep it with one emulator running right now, the fastest non-flaky one (somewhere in the middle of our range)

Then I'll peel an issue from this comment where I dynamically build a matrix, and set it to run on a daily to do a matrix that's actually a low/high bracket, but is not as sensitive to emulator performance - https://stackoverflow.com/questions/65384420/how-to-make-a-github-action-matrix-element-conditional

Also I'll peel an issue to switch our unit test gradle cache strategy to the one used here (read--only on branch, read/write on main) as I think that will reduce cache eviction and speed aggregate time for all builds

**Comment by mikehardy** [comment]
Okay, this is as good as it can be - I can't move further on it until it's merged to main and I can start observing the cache read/write behavior interaction between main branch and PR branches. I'll be :eagle: :eyes: on it though, there's a good optimization here to be had from optimizing LRU cache behavior for gradle and AVD snapshot
